File: src/de/hsh/dennis/model/NpcLogic/actors/Bot.java
package de.hsh.dennis.model.NpcLogic.actors;


import de.hsh.dennis.model.NpcLogic.SkinConfig;
import de.hsh.dennis.model.NpcLogic.NPCEnums;

import java.io.FileNotFoundException;

public class Bot extends Npc {

    private static final String picturePath = "/dennis/skins/bot/bot.png";
    private static final int defaultSpeed = 1;

    private double bounceMax = 5.0;
    private double bouncePos = getBounceMax() * -1.0;
    private double bounceSpeed = getSpeed() / 2.0;
    private boolean bounceInitialized = false;


    public Bot(NPCEnums.Spawn spawnType) throws FileNotFoundException {
        super(SkinConfig.Bot.skin_standard_path, spawnType, NPCEnums.NpcType.BOT);
        setCurrentImage(SkinConfig.Bot.skin_standard);
        bounceInit();
    }

    public Bot(NPCEnums.Spawn spawnType, double spawnTime) throws FileNotFoundException {
        super(SkinConfig.Bot.skin_standard_path, spawnType, NPCEnums.NpcType.BOT, spawnTime, defaultSpeed);
        setCurrentImage(SkinConfig.Bot.skin_standard);

        bounceInit();
    }

    public Bot(NPCEnums.Spawn spawnType, double spawnTime, double speed) throws FileNotFoundException {
        super(SkinConfig.Bot.skin_standard_path, spawnType, NPCEnums.NpcType.BOT, spawnTime, speed);
        setCurrentImage(SkinConfig.Bot.skin_standard);

        bounceInit();
    }

    @Override
    public void move() {
        setPosX((getSpawnType() == NPCEnums.Spawn.RIGHT) ? (getPosX() - getSpeed()) : (getPosX() + getSpeed()));
        bounce();

    }

    private void bounceInit() {
        if (!bounceInitialized) {
            setPosY(getPosY() - getBounceMax());
            bounceInitialized = !bounceInitialized;
        }
    }

    private void bounce() {


        if (bounceMax > 0.0) {
            if (bouncePos < getBounceMax()) {
                if ((bouncePos + bounceSpeed) > getBounceMax()) {
                    bouncePos = getBounceMax();
                } else {
                    bouncePos += bounceSpeed;
                }
            } else if (bouncePos >= getBounceMax()) {
                setBounceMax(getBounceMax() * -1.0);
            }
        } else {
            if (bouncePos > getBounceMax()) {
                if ((bouncePos - bounceSpeed) < getBounceMax()) {
                    bouncePos = getBounceMax();
                } else {
                    bouncePos -= bounceSpeed;
                }
            } else if (bouncePos <= getBounceMax()) {
                setBounceMax(getBounceMax() * -1.0);
            }
        }
        setPosY(getPosY() + bouncePos);
    }

    public double getBounceMax() {
        return bounceMax;
    }

    public void setBounceMax(double bounceMax) {
        this.bounceMax = bounceMax;
    }
}

File: Games/Dennis/src/de/hsh/dennis/model/Spawn/actors/Bot.kt
package de.hsh.dennis.model.Spawn.actors

import common.actor.Direction
import common.config.WindowConfig
import de.hsh.dennis.model.Spawn.SkinConfig
import de.hsh.dennis.model.Spawn.Spawn

class Bot(spawnType: Spawn, spawnTime: Double, speed: Double) :
        Npc(SkinConfig.Bot.skin_standard_path,
            spawnType,
            spawnTime,
            speed) {

    var bounceMax = 5.0
    private var bouncePos: Double = bounceMax * -1.0
    private val bounceSpeed: Double = speed / 2.0

    init {
        x = if (spawnType == Spawn.RIGHT) WindowConfig.window_width.toDouble()
        else -SkinConfig.Bot.skin_width.toDouble()
        y = SkinConfig.Player.posY + SkinConfig.Player.skin_height * 0.2 - bounceMax
    }

    override fun collidesWithPlayer(player: Player): Boolean =
            (this.spawnType == Spawn.RIGHT && player.direction === Direction.Right ||
             this.spawnType == Spawn.LEFT && player.direction === Direction.Left) &&
            player.doesCollide(this)

    override fun move() {
        super.move()
        bounce()
    }

    private fun bounce() {
        if (bounceMax > 0.0) {
            if (bouncePos < bounceMax) {
                if (bouncePos + bounceSpeed > bounceMax) bouncePos = bounceMax
                else bouncePos += bounceSpeed
            } else {
                bounceMax *= -1.0
            }
        } else {
            if (bouncePos > bounceMax) {
                if (bouncePos - bounceSpeed < bounceMax) bouncePos = bounceMax
                else bouncePos -= bounceSpeed
            } else {
                bounceMax *= -1.0
            }
        }
        y += bouncePos
    }

}
